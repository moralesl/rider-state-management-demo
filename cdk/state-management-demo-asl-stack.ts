import * as fs from "fs";
import * as path from "path";

import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import * as dynamodb from "aws-cdk-lib/aws-dynamodb";
import * as lambda from "aws-cdk-lib/aws-lambda";
import * as sfn from "aws-cdk-lib/aws-stepfunctions";
import * as sns from "aws-cdk-lib/aws-sns";
import * as sqs from "aws-cdk-lib/aws-sqs";
import { Duration, RemovalPolicy } from "aws-cdk-lib";

export class StateManagementASLDemoStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const riderStateValidation = new lambda.Function(this, "RiderStateValidation-ASL-CDK", {
      functionName: "RiderStateValidation-ASL-CDK",
      runtime: lambda.Runtime.PYTHON_3_9,
      architecture: lambda.Architecture.ARM_64,
      handler: "app.handler",
      code: lambda.Code.fromAsset(path.join(__dirname, "..", "rider-state-validation")),
    });

    const riderStateChangeEvent = new sns.Topic(this, "RiderStateChangeEvent-ASL-CDK", {
      displayName: "RiderStateChangeEvent-ASL-CDK",
      topicName: "RiderStateChangeEvent-ASL-CDK"
    });

    const riderStateDLQ = new sqs.Queue(this, "RiderStateDLQ-ASL-CDK", {
      queueName: "RiderStateDLQ-ASL-CDK",
      retentionPeriod: Duration.minutes(5),

      removalPolicy: RemovalPolicy.DESTROY, // This is just for development purposes
    });

    const riderStateTable = new dynamodb.Table(this, "RiderStateTable-ASL-CDK", {
      partitionKey: { name: "Area#Entity", type: dynamodb.AttributeType.STRING },
      billingMode: dynamodb.BillingMode.PAY_PER_REQUEST,
      tableName: "RiderStateTable-ASL-CDK",

      removalPolicy: RemovalPolicy.DESTROY, // This is just for development purposes
    });

    riderStateTable.addGlobalSecondaryIndex({
      indexName: "State",
      partitionKey: { name: "State", type: dynamodb.AttributeType.STRING },
      sortKey: { name: "Area#Entity", type: dynamodb.AttributeType.STRING },

      projectionType: dynamodb.ProjectionType.ALL,
    });

    const riderStateTransitionManagementStateMachine = new sfn.StateMachine(
      this,
      "RiderStateTransitionManagement-ASL-CDK",
      {
        definition: new sfn.Pass(this, "StartState"),
        tracingEnabled: true,
        stateMachineName: "RiderStateTransitionManagement-ASL-CDK",
        stateMachineType: sfn.StateMachineType.EXPRESS,
      }
    );

    const cfnStatemachine = riderStateTransitionManagementStateMachine.node.defaultChild as sfn.CfnStateMachine;

    const stateMachineDefinition = JSON.parse(fs.readFileSync("rider-state-management.asl.json", "utf8"));
    cfnStatemachine.definitionString = JSON.stringify(stateMachineDefinition);

    riderStateValidation.grantInvoke(riderStateTransitionManagementStateMachine);
    riderStateChangeEvent.grantPublish(riderStateTransitionManagementStateMachine);
    riderStateDLQ.grantSendMessages(riderStateTransitionManagementStateMachine);
    riderStateTable.grantReadWriteData(riderStateTransitionManagementStateMachine);
  }
}
