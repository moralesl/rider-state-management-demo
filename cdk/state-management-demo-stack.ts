import * as path from "path";

import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import * as dynamodb from "aws-cdk-lib/aws-dynamodb";
import * as lambda from "aws-cdk-lib/aws-lambda";
import * as logs from "aws-cdk-lib/aws-logs";
import * as sfn from "aws-cdk-lib/aws-stepfunctions";
import { Choice, Condition, JsonPath } from "aws-cdk-lib/aws-stepfunctions";
import * as tasks from "aws-cdk-lib/aws-stepfunctions-tasks";
import * as sns from "aws-cdk-lib/aws-sns";
import * as sqs from "aws-cdk-lib/aws-sqs";
import { CfnOutput, Duration, RemovalPolicy } from "aws-cdk-lib";

export class StateManagementDemoStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const riderStateTable = new dynamodb.Table(this, "RiderStateTable-CDK", {
      partitionKey: { name: "Area#Entity", type: dynamodb.AttributeType.STRING },
      billingMode: dynamodb.BillingMode.PAY_PER_REQUEST,
      tableName: "RiderStateTable-CDK",

      removalPolicy: RemovalPolicy.DESTROY, // This is just for development purposes
    });

    riderStateTable.addGlobalSecondaryIndex({
      indexName: "State",
      partitionKey: { name: "State", type: dynamodb.AttributeType.STRING },
      sortKey: { name: "Area#Entity", type: dynamodb.AttributeType.STRING },

      projectionType: dynamodb.ProjectionType.ALL,
    });

    // 1. Get current rider state informations
    const getCurrentRiderStateInformation = new tasks.DynamoGetItem(this, "Get current rider state", {
      table: riderStateTable,
      key: {
        "Area#Entity": tasks.DynamoAttributeValue.fromString(JsonPath.stringAt("$.input.rider_id")),
      },
      resultSelector: {
        "rider_id.$": "$.Item.Area#Entity.S",
        "timestamp.$": "$.Item.Timestamp.N",
        "state.$": "$.Item.State.S",
        "lat.$": "$.Item.Lat.N",
        "long.$": "$.Item.Long.N",
      },
      resultPath: "$.rider",
    }).addRetry({
      maxAttempts: 3,
      interval: Duration.seconds(1),
      backoffRate: 1.5,
    });

    // 2. Validate starting point
    const validateStartPoint = new tasks.LambdaInvoke(this, "Validate start point and device token", {
      lambdaFunction: new lambda.Function(this, "RiderStateValidation-CDK", {
        functionName: "RiderStateValidation-CDK",
        runtime: lambda.Runtime.PYTHON_3_9,
        architecture: lambda.Architecture.ARM_64,
        handler: "app.handler",
        code: lambda.Code.fromAsset(path.join(__dirname, "..", "rider-state-validation")),
      }),
      payload: sfn.TaskInput.fromJsonPathAt("$"),
      resultSelector: {
        "statusCode.$": "$.Payload.statusCode",
        "valid.$": "$.Payload.valid",
      },
      resultPath: "$.validator",
    }).addRetry({
      maxAttempts: 5,
      interval: Duration.seconds(1),
      backoffRate: 1.5,
    });

    // 3. Transition rider to next state & persist it
    const transitionRiderToNextStateAndPersistIt = new tasks.DynamoUpdateItem(this, "Transition rider to next state", {
      table: riderStateTable,
      key: {
        "Area#Entity": tasks.DynamoAttributeValue.fromString(JsonPath.stringAt("$.input.rider_id")),
      },
      updateExpression: "SET #state = :nextState",
      expressionAttributeNames: {
        "#state": "State",
      },
      expressionAttributeValues: {
        ":nextState": tasks.DynamoAttributeValue.fromString(JsonPath.stringAt("$.input.next_state")),
      },
      resultSelector: {
        "status_code.$": "$.SdkHttpMetadata.HttpStatusCode",
      },
      resultPath: "$.ddb",
    }).addRetry({
      maxAttempts: 5,
      interval: Duration.seconds(1),
      backoffRate: 1.5,
    });

    // 4. Emit rider state change information
    const emitRiderStateChangeEvent = new tasks.SnsPublish(this, "Emit rider state change event", {
      topic: new sns.Topic(this, "RiderStateChangeEvent-CDK", {
        displayName: "RiderStateChangeEvent-CDK",
        topicName: "RiderStateChangeEvent-CDK",
      }),
      message: sfn.TaskInput.fromJsonPathAt("$"),
      resultSelector: {
        "status_code.$": "$.SdkHttpMetadata.HttpStatusCode",
      },
      resultPath: "$.sns",
    }).addRetry({
      maxAttempts: 5,
      interval: Duration.seconds(1),
      backoffRate: 1.5,
    });

    // On error send event to DLQ
    const sendEventToDlqForManualHandling = new tasks.SqsSendMessage(this, "Send event to DLQ for manual handling", {
      queue: new sqs.Queue(this, "RiderStateDLQ-CDK", {
        queueName: "RiderStateDLQ-CDK",
        retentionPeriod: Duration.minutes(5),

        removalPolicy: RemovalPolicy.DESTROY, // This is just for development purposes
      }),
      messageBody: sfn.TaskInput.fromJsonPathAt("$"),
    })
      .addRetry({
        maxAttempts: 5,
        interval: Duration.seconds(1),
        backoffRate: 1.5,
      })
      .next(new sfn.Fail(this, "State Transition Failed", {}));

    const riderStateTransitionManagementStateMachineDefinition = new Choice(this, "Is rider id and next state valid?")
      .when(
        // Check if input is valid
        Condition.and(
          ...allArePresent("$.input.rider_id", "$.input.next_state"),
          anyStringMatches("$.input.next_state", "Not Working", "Available", "Starting", "Working")
        ),
        // Get current rider state information
        getCurrentRiderStateInformation.next(
          new Choice(this, "Is next state a valid transition?")
            .when(
              // If it transitions from starting to working...
              isValidStateTransition({
                currentStateVariable: "$.rider.state",
                nextStateVariable: "$.input.next_state",
                validTransitions: [["Starting", "Working"]],
              }),
              // Validate the start point and transition the rider to the next state
              validateStartPoint.next(transitionRiderToNextStateAndPersistIt)
            )
            .when(
              // If it is another valid transition...
              isValidStateTransition({
                currentStateVariable: "$.rider.state",
                nextStateVariable: "$.input.next_state",
                validTransitions: [
                  ["Not Working", "Available"],
                  ["Not Working", "Starting"],
                  ["Available", "Not Working"],
                  ["Available", "Starting"],
                  ["Starting", "Not Working"],
                  ["Working", "Not Working"],
                ],
              }),
              // Transition the rider to the next state and emit change event
              transitionRiderToNextStateAndPersistIt.next(emitRiderStateChangeEvent)
            )
            // Otherwise send event to DLQ
            .otherwise(sendEventToDlqForManualHandling)
        )
      )
      // Otherwise send event to DLQ
      .otherwise(sendEventToDlqForManualHandling);

    const riderStateTransitionManagementStateMachine = new sfn.StateMachine(
      this,
      "RiderStateTransitionManagement-CDK",
      {
        definitionBody: sfn.DefinitionBody.fromChainable(riderStateTransitionManagementStateMachineDefinition),
        stateMachineName: "RiderStateTransitionManagement-CDK",
        stateMachineType: sfn.StateMachineType.EXPRESS,
        tracingEnabled: true,
        logs: {
          destination: new logs.LogGroup(this, "RiderStateTransitionManagement-CDK-LogGroup"),
          level: sfn.LogLevel.ALL,
          includeExecutionData: true,
        },
      }
    );

    riderStateTable.grantReadWriteData(riderStateTransitionManagementStateMachine);

    new CfnOutput(this, "RiderStateTransitionManagementStateMachine-Arn", {
      value: riderStateTransitionManagementStateMachine.stateMachineArn,
    });
  }
}

const allArePresent = (...fieldsToBePresent: string[]): Condition[] => {
  return fieldsToBePresent.map((fieldToBePresent) => Condition.isPresent(fieldToBePresent));
};

const anyStringMatches = (fieldToCheck: string, ...possibleMatches: string[]): Condition => {
  return Condition.or(...possibleMatches.map((possibleMatch) => Condition.stringEquals(fieldToCheck, possibleMatch)));
};

const isValidStateTransition = (config: {
  currentStateVariable: string;
  nextStateVariable: string;
  validTransitions: string[][];
}): Condition => {
  return Condition.or(
    ...config.validTransitions.map((validTransition) =>
      Condition.and(
        Condition.stringEquals(config.currentStateVariable, validTransition[0]),
        Condition.stringEquals(config.nextStateVariable, validTransition[1])
      )
    )
  );
};
