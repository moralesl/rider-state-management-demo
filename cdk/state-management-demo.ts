#!/usr/bin/env node
import 'source-map-support/register';
import * as cdk from 'aws-cdk-lib';
import { StateManagementASLDemoStack } from './state-management-demo-asl-stack';
import { StateManagementDemoStack } from './state-management-demo-stack';
import { Tags } from 'aws-cdk-lib';

const app = new cdk.App();
const stateManagementASLDemoStack = new StateManagementASLDemoStack(app, 'StateManagementASLDemoStack', {
  env: { account: process.env.CDK_DEFAULT_ACCOUNT, region: 'us-east-1'},
});

Tags.of(stateManagementASLDemoStack).add('Project', 'StateManagementASLDemo')


const stateManagementDemoStack = new StateManagementDemoStack(app, 'StateManagementDemoStack', {
  env: { account: process.env.CDK_DEFAULT_ACCOUNT, region: 'us-east-1'},
});

Tags.of(stateManagementDemoStack).add('Project', 'StateManagementDemo')
