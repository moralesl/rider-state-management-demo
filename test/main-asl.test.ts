import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { StateManagementASLDemoStack } from '../cdk/state-management-demo-asl-stack';

test('ASL Demo match Snapshot', () => {
  const app = new App();
  const stack = new StateManagementASLDemoStack(app, 'StateManagementASLDemoStack');

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});
