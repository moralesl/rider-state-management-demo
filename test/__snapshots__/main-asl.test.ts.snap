// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ASL Demo match Snapshot 1`] = `
Object {
  "Mappings": Object {
    "ServiceprincipalMap": Object {
      "af-south-1": Object {
        "states": "states.af-south-1.amazonaws.com",
      },
      "ap-east-1": Object {
        "states": "states.ap-east-1.amazonaws.com",
      },
      "ap-northeast-1": Object {
        "states": "states.ap-northeast-1.amazonaws.com",
      },
      "ap-northeast-2": Object {
        "states": "states.ap-northeast-2.amazonaws.com",
      },
      "ap-northeast-3": Object {
        "states": "states.ap-northeast-3.amazonaws.com",
      },
      "ap-south-1": Object {
        "states": "states.ap-south-1.amazonaws.com",
      },
      "ap-south-2": Object {
        "states": "states.ap-south-2.amazonaws.com",
      },
      "ap-southeast-1": Object {
        "states": "states.ap-southeast-1.amazonaws.com",
      },
      "ap-southeast-2": Object {
        "states": "states.ap-southeast-2.amazonaws.com",
      },
      "ap-southeast-3": Object {
        "states": "states.ap-southeast-3.amazonaws.com",
      },
      "ap-southeast-4": Object {
        "states": "states.ap-southeast-4.amazonaws.com",
      },
      "ca-central-1": Object {
        "states": "states.ca-central-1.amazonaws.com",
      },
      "cn-north-1": Object {
        "states": "states.cn-north-1.amazonaws.com",
      },
      "cn-northwest-1": Object {
        "states": "states.cn-northwest-1.amazonaws.com",
      },
      "eu-central-1": Object {
        "states": "states.eu-central-1.amazonaws.com",
      },
      "eu-central-2": Object {
        "states": "states.eu-central-2.amazonaws.com",
      },
      "eu-north-1": Object {
        "states": "states.eu-north-1.amazonaws.com",
      },
      "eu-south-1": Object {
        "states": "states.eu-south-1.amazonaws.com",
      },
      "eu-south-2": Object {
        "states": "states.eu-south-2.amazonaws.com",
      },
      "eu-west-1": Object {
        "states": "states.eu-west-1.amazonaws.com",
      },
      "eu-west-2": Object {
        "states": "states.eu-west-2.amazonaws.com",
      },
      "eu-west-3": Object {
        "states": "states.eu-west-3.amazonaws.com",
      },
      "il-central-1": Object {
        "states": "states.il-central-1.amazonaws.com",
      },
      "me-central-1": Object {
        "states": "states.me-central-1.amazonaws.com",
      },
      "me-south-1": Object {
        "states": "states.me-south-1.amazonaws.com",
      },
      "sa-east-1": Object {
        "states": "states.sa-east-1.amazonaws.com",
      },
      "us-east-1": Object {
        "states": "states.us-east-1.amazonaws.com",
      },
      "us-east-2": Object {
        "states": "states.us-east-2.amazonaws.com",
      },
      "us-gov-east-1": Object {
        "states": "states.us-gov-east-1.amazonaws.com",
      },
      "us-gov-west-1": Object {
        "states": "states.us-gov-west-1.amazonaws.com",
      },
      "us-iso-east-1": Object {
        "states": "states.amazonaws.com",
      },
      "us-iso-west-1": Object {
        "states": "states.amazonaws.com",
      },
      "us-isob-east-1": Object {
        "states": "states.amazonaws.com",
      },
      "us-west-1": Object {
        "states": "states.us-west-1.amazonaws.com",
      },
      "us-west-2": Object {
        "states": "states.us-west-2.amazonaws.com",
      },
    },
  },
  "Outputs": Object {
    "RiderStateTransitionManagementStateMachineASLArn": Object {
      "Value": Object {
        "Ref": "RiderStateTransitionManagementASLCDK137D569C",
      },
    },
  },
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "RiderStateChangeEventASLCDK464CBE26": Object {
      "Properties": Object {
        "DisplayName": "RiderStateChangeEvent-ASL-CDK",
        "TopicName": "RiderStateChangeEvent-ASL-CDK",
      },
      "Type": "AWS::SNS::Topic",
    },
    "RiderStateDLQASLCDKB605F866": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "MessageRetentionPeriod": 300,
        "QueueName": "RiderStateDLQ-ASL-CDK",
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "RiderStateTableASLCDK34375312": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "Area#Entity",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "State",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "GlobalSecondaryIndexes": Array [
          Object {
            "IndexName": "State",
            "KeySchema": Array [
              Object {
                "AttributeName": "State",
                "KeyType": "HASH",
              },
              Object {
                "AttributeName": "Area#Entity",
                "KeyType": "RANGE",
              },
            ],
            "Projection": Object {
              "ProjectionType": "ALL",
            },
          },
        ],
        "KeySchema": Array [
          Object {
            "AttributeName": "Area#Entity",
            "KeyType": "HASH",
          },
        ],
        "TableName": "RiderStateTable-ASL-CDK",
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
    "RiderStateTransitionManagementASLCDK137D569C": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "RiderStateTransitionManagementASLCDKRoleDefaultPolicyF78997AA",
        "RiderStateTransitionManagementASLCDKRole37A1FC67",
      ],
      "Properties": Object {
        "DefinitionS3Location": Object {
          "Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "Key": "d842334aea3ffd1b5252dc2f22d62d7eb4f7c68cfd8d74451edf4f5b06d2f585.json",
        },
        "LoggingConfiguration": Object {
          "Destinations": Array [
            Object {
              "CloudWatchLogsLogGroup": Object {
                "LogGroupArn": Object {
                  "Fn::GetAtt": Array [
                    "RiderStateTransitionManagementASLCDKLogGroupA25010EE",
                    "Arn",
                  ],
                },
              },
            },
          ],
          "IncludeExecutionData": true,
          "Level": "ALL",
        },
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "RiderStateTransitionManagementASLCDKRole37A1FC67",
            "Arn",
          ],
        },
        "StateMachineName": "RiderStateTransitionManagement-ASL-CDK",
        "StateMachineType": "EXPRESS",
        "TracingConfiguration": Object {
          "Enabled": true,
        },
      },
      "Type": "AWS::StepFunctions::StateMachine",
      "UpdateReplacePolicy": "Delete",
    },
    "RiderStateTransitionManagementASLCDKLogGroupA25010EE": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "RetentionInDays": 731,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "RiderStateTransitionManagementASLCDKRole37A1FC67": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::FindInMap": Array [
                    "ServiceprincipalMap",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    "states",
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "RiderStateTransitionManagementASLCDKRoleDefaultPolicyF78997AA": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogDelivery",
                "logs:GetLogDelivery",
                "logs:UpdateLogDelivery",
                "logs:DeleteLogDelivery",
                "logs:ListLogDeliveries",
                "logs:PutResourcePolicy",
                "logs:DescribeResourcePolicies",
                "logs:DescribeLogGroups",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
                "xray:GetSamplingRules",
                "xray:GetSamplingTargets",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "RiderStateValidationASLCDK21EFCC6C",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "RiderStateValidationASLCDK21EFCC6C",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": Object {
                "Ref": "RiderStateChangeEventASLCDK464CBE26",
              },
            },
            Object {
              "Action": Array [
                "sqs:SendMessage",
                "sqs:GetQueueAttributes",
                "sqs:GetQueueUrl",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "RiderStateDLQASLCDKB605F866",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": Array [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "RiderStateTableASLCDK34375312",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "RiderStateTableASLCDK34375312",
                          "Arn",
                        ],
                      },
                      "/index/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "RiderStateTransitionManagementASLCDKRoleDefaultPolicyF78997AA",
        "Roles": Array [
          Object {
            "Ref": "RiderStateTransitionManagementASLCDKRole37A1FC67",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "RiderStateValidationASLCDK21EFCC6C": Object {
      "DependsOn": Array [
        "RiderStateValidationASLCDKServiceRoleC9BA78D9",
      ],
      "Properties": Object {
        "Architectures": Array [
          "arm64",
        ],
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "7751bbd3241187cb72efbcbc46253e7f24e94542772c8348f6ef4f0008688485.zip",
        },
        "FunctionName": "RiderStateValidation-ASL-CDK",
        "Handler": "app.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "RiderStateValidationASLCDKServiceRoleC9BA78D9",
            "Arn",
          ],
        },
        "Runtime": "python3.9",
      },
      "Type": "AWS::Lambda::Function",
    },
    "RiderStateValidationASLCDKServiceRoleC9BA78D9": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
